llmfoundry/optim/lion8b.py

```diff
@@ -0,0 +1,417 @@
+                    qtensor.state_dict(
+                        name=_KEY_MOMENTUM,
+                        allow_quantized=self._compress_state_dict))
+            opt_state[param_id] = param_state
+        return d
+
+
+class _MaybeQuantizedTensor:
+    """Helper class so 8b LION doesn't have to know quantization details.
+
+    Important points about this class:
+    * It handles CPU tensors not being quantized
+    * It knows how to save + load state dicts, handling both the quantized
+        and not quantized cases
+    * It implements some parts of the torch.Tensor interface that we need,
+        but is not intended to be a full torch.Tensor replacement
+    """
+
+    def __init__(self, data: Optional[torch.Tensor], try_quantize: bool = True):
+        super().__init__()
+        self.data: Optional[torch.Tensor] = None
+        self.quantized: Optional[torch.Tensor] = None
+        self.scales: Optional[torch.Tensor] = None
+        self._try_quantize = try_quantize and torch.cuda.is_available()
+
+        # conditionally import CUDA kernels
+        self._f_encode = None
+        self._f_decode = None
+        if self._try_quantize:
+            from turbo import dequantize8b, quantize8b
+            self._f_encode = quantize8b
+            self._f_decode = dequantize8b
+
+        if data is not None:
+            self.set_data(data)
+
+    def state_dict(self,
+                   name: str,
+                   allow_quantized: bool = False) -> Dict[str, torch.Tensor]:
+        if self.is_quantized() and allow_quantized:
+            assert self.quantized is not None  # pyright
+            assert self.scales is not None  # pyright
+            return {
+                f'{name}::quantized': self.quantized,
+                f'{name}::scales': self.scales
+            }
+        return {name: self.materialize().to(dtype=torch.bfloat16)}
+
+    def load_state_dict(self, d: Dict[str, torch.Tensor], name: str) -> None:
+        # we allow other keys in the state dict for convenience, so you can
+        # just pass this the whole opt state for a parameters
+        d = {k: v for k, v in d.items() if k.startswith(name)}
+        if name in d:
+            if len(d) != 1:
+                raise ValueError(
+                    f'If state dict specifies {name}, it must not ' +
+                    f'specify other keys. Got {list(d.keys())}')
+            self.set_data(d[name])
+            return
+
+        self.quantized = d[f'{name}::quantized'].to(dtype=torch.int8)
+        self.scales = d[f'{name}::scales'].to(dtype=torch.float16)
+
+    def set_data(self, data: torch.Tensor) -> None:
+        if not (self._try_quantize and data.is_cuda):
+            self.data = data.to(dtype=torch.float32)
+            self.quantized = None
+            self.scales = None
+        else:
+            self.data = None
+            assert self._f_encode is not None  # pyright
+            self.quantized, self.scales = self._f_encode(data)
```

Comment:
flipping the if else makes more sense (the `not` can throw ppl off)

Should an error be thrown if `self._try_quantize and not data.is_cuda`?

Reply:
Flipped the if and else and made this throw.